src/HooglePlus/FilterTest.hs

```diff
@@ -0,0 +1,194 @@
+
+    liftIO act
+  }
+  case result of
+    Left err -> putStrLn (displayException err) >> return "error \"114514\""
+    Right res -> return ("(" ++ res ++ ")")
+
+mapRandomMultiple :: (TypeEnv, [ArgumentType], ArgumentType) -> Int -> IO [String]
+mapRandomMultiple env count = do
+  base <- mapRandomArguments' env
+  replicateM count (derive base)
+  where
+    derive :: [(String, String)] -> IO String
+    derive base = do
+      rand <- generate (arbitrary :: Gen Int)
+      let index = rand `rem` length base
+      let item = getNthElement index base
+
+      replaced <- newRandomValue item
+      let list = setNthElement index replaced base
+      return $ unwords (map snd list)
+
+    newRandomValue (name, _) = do
+      val <- generateRandomValue name
+      return (name, val)
+
+    getNthElement idx list = item
+      where (_, item:_) = splitAt idx list
+
+    setNthElement idx item list
+      | idx < 0   = list
+      | otherwise = case splitAt idx list of
+                    (front, _:back) -> front ++ item : back
+                    _ -> list
+
+mapRandomArguments :: (TypeEnv, [ArgumentType], ArgumentType) -> IO String
+mapRandomArguments env = do
+  result <- mapRandomArguments' env
+  return $ unwords (map snd result)
+
+mapRandomArguments' :: (TypeEnv, [ArgumentType], ArgumentType) -> IO [(String, String)]
+mapRandomArguments' (typeEnv, args, retType) = mapM f (transformType typeEnv args)
+  where
+    f x = let str = show x in do
+      val <- generateRandomValue str
+      return (str, val)
+
+    transformType [] args = args
+    transformType (envItem:env) args = transformType env args'
+      where
+        args' = map (replace envItem) args
+
+        replace (Polymorphic name, cons) (Polymorphic name') =
+          if name == name' then
+            case cons of
+              "Eq" -> Concrete "Int"
+              "Num" -> Concrete "Int"
+          else Polymorphic name'
+        replace cons (ArgTypeList argType) = ArgTypeList (replace cons argType)
+        replace _ x = x
+
+eval_ :: [String] -> String -> Int -> IO (Maybe (Either InterpreterError String))
+eval_ modules line time = timeout time $ runInterpreter $ do {
+  setImports modules;
+  eval line
+}
+
+runChecks :: Environment -> RType -> UProgram -> IO Bool
+runChecks env goalType prog = do
+  list <- replicateM 10 (runChecks' modules funcSig body)
+  return $ or list
+  where
+    args = _arguments env
+    modules = "Prelude" : Set.toList (_included_modules env)
+    argList = Map.toList args
+    argNames = map fst argList
+    argTypes = map snd argList
+    monoGoals = map toMonotype argTypes
+    funcSig = mkFunctionSigStr (monoGoals ++ [goalType])
+    body = mkLambdaStr argNames prog
+
+runChecks' :: [String] -> String -> String -> IO Bool
+runChecks' modules funcSig body = if filterInf body then handleNotSupported $ do
+  arg <- (mapRandomArguments . parseTypeString) funcSig
+  result <- eval_ modules (fmtFunction_ body funcSig arg) 3000000
```

Comment:
You should lift the timeout to a variable and colocate it with the types and default values.

Reply:
Could you please elaborate more on that? I'm not sure how to lift the timeout to a variable.